src/Control/Bind.purs

```diff
@@ -66,6 +68,17 @@ instance bindArray :: Bind Array where
 
 foreign import arrayBind :: forall a b. Array a -> (a -> Array b) -> Array b
 
+-- | A class for types whose values can safely be discarded
+-- | in a `do` notation block.
+-- |
+-- | An example is the `Unit` type, since there is only one
+-- | possible value which can be returned.
+class Discard a where
+  discard :: forall f b. Apply f => f a -> (a -> f b) -> f b
+
+instance discardUnit :: Discard Unit where
+  discard u f = u *> f unit
```

Comment:
Maybe this needs to use `>>=` actually, if we don't want it to always strictly evaluate `f`.